File: training/objectivec/CellarsAndCentaurs/Domain/Model/HitPoints.m
#import "HitPoints.h"

@implementation HitPoints

- (instancetype)initWithCurrent:(NSInteger)current max:(NSInteger)max temporary:(NSInteger)temporary {
    self = [super init];
    if (self) {
        _current = MIN(current, MAX_HIT_POINTS);
        _max = MIN(max, MAX_HIT_POINTS);
        _temporary = temporary;
    }
    return self;
}

+ (instancetype)hitPointsWithMax:(NSInteger)max {
    return [[HitPoints alloc] initWithCurrent:max max:max temporary:0];
}

- (void)loseHitPoints:(NSInteger)damage {
    self.current = MAX(0, self.current - damage);
}

@end

File: training/vue/src/de/sots/cellarsandcentaurs/domain/model/HitPoints.ts
/**
 * Hit points drop when the creature takes damage and recover when it rests in its lair.
 */
export class HitPoints {
  static readonly MAX_HIT_POINTS = 999

  constructor(
    readonly current: number,
    readonly max: number,
    readonly temporary: number = 0
  ) {}

  static init(max: number): HitPoints {
    return new HitPoints(Math.min(max, HitPoints.MAX_HIT_POINTS), max, 0)
  }

  takeDamage(amount: number): HitPoints {
    return new HitPoints(Math.max(0, this.current - amount), this.max, this.temporary)
  }

  rest(): HitPoints {
    return new HitPoints(this.max, this.max, 0)
  }
}
